import "common.proto";
import "type.proto";
import "tablet.proto";
package rtidb.blobserver;

option cc_generic_services = true;
option java_generic_services = true;
option java_package = "com._4paradigm.rtidb.object_storage_server";
option java_outer_classname = "ObjectStorage";

message GeneralResponse {
    optional string msg = 1;
    optional int32 code = 2;
};

message GeneralRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
    optional string key = 3;
    optional bool attachment = 4 [default = false];
}

message GetResponse {
    optional bytes pairs = 1;
    optional string msg = 2;
    optional int32 code = 3;
}

message PutRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
    optional string key = 3;
    optional bytes data = 4;
}

message PutResponse {
    optional uint32 code = 1;
    optional string msg = 2;
    optional string key = 3;
}

message TableMeta {
    optional uint32 tid = 1;
    optional string name = 2;
    optional uint32 pid = 3;
    optional rtidb.common.StorageMode storage_mode = 4 [default = kHDD];
    optional rtidb.type.TableType table_type = 5;
    optional rtidb.api.CompressType compress_type = 6;
    optional uint32 height = 7 [default = 1];
    optional int64 before_time = 8 [default = 0];
}

message CreateTableRequest {
    optional TableMeta table_meta = 1;
}

message StatsRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
}

message StatsResponse {
    optional string msg = 1;
    optional int32 code = 2;
    optional uint64 count = 3;
    optional uint64 total_space = 4;
    optional uint64 avail_space = 5;
}

message LoadTableRequest {
    optional TableMeta table_meta = 1;
    optional rtidb.api.TaskInfo task_info = 2;
}

message StoreStatus {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
}

message GetStoreStatusResponse {
    repeated StoreStatus all_status = 1;
    optional int32 code = 2;
    optional string msg = 3;
}

message GetStoreStatusRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
}


service BlobServer {
    rpc CreateTable(CreateTableRequest) returns (GeneralResponse);
    rpc Put(PutRequest) returns (PutResponse);
    rpc Get(GeneralRequest) returns (GetResponse);
    rpc Delete(GeneralRequest) returns (GeneralResponse);
    rpc Stats(StatsRequest) returns (StatsResponse);
    rpc LoadTable(LoadTableRequest) returns (GeneralResponse);
    rpc GetStoreStatus(GetStoreStatusRequest) returns (GetStoreStatusResponse);
}
